### PR TITLE
feat(secrets): add vault secrets support

### PR DESCRIPTION
**Pull request by KathrynLewis**

### CODE REVIEW

File: pkg/secrets/vault.go [hunk 1]
```diff
@@ -2,23 +2,242 @@ package secrets
...
-	return &VaultDecrypter{secretConfig}
+func getVaultDecrypter(encryptedSecret string) Decrypter {
+	decrypter := NewVaultDecrypter(encryptedSecret)
+	decrypter.vaultConfig = Registry.VaultConfig
+	return decrypter
 }
 
 func (v *VaultDecrypter) Decrypt() (string, error) {
-	return "", fmt.Errorf("vault secrets unsupported")
+	if err := v.ValidateVaultConfig(); err != nil {
```
Comment (ncknt) [inline pkg/secrets/vault.go]: We shouldn't validate while decrypting (so for every secret) but when the config is set.

File: pkg/secrets/vault.go [hunk 2]
```diff
@@ -2,23 +2,242 @@ package secrets
...
+	if v.vaultConfig.Token == "" {
+		token, err := v.FetchVaultToken()
+		if err != nil {
+			return "", fmt.Errorf("error fetching vault token - %s", err)
+		}
+		v.vaultConfig.Token = token
+	}
+
+	secret, err := v.FetchSecret(vaultSecret)
+	if err != nil {
```
Comment (ncknt) [inline pkg/secrets/vault.go]: should we check the type of error before retrying? It should be not found, not 403 etc..

File: pkg/secrets/vault.go [hunk 3]
```diff
@@ -2,23 +2,242 @@ package secrets
...
+		if token != "" {
+			return token, nil
+		} else {
+			return "", fmt.Errorf("VAULT_TOKEN environment variable not set")
+		}
+	} else if v.vaultConfig.AuthMethod == "KUBERNETES" {
+		if v.vaultConfig.Path == "" || v.vaultConfig.Role == "" {
+			return "", fmt.Errorf("path and role both required for Kubernetes auth method")
+		}
+		return v.FetchServiceAccountToken()
```
Comment (ncknt) [inline pkg/secrets/vault.go]: Is there no Vault SDK method to handle kubernetes service account login? That would be surprising.

File: pkg/secrets/vault.go (hunk 1 above)
Comment (KathrynLewis) [reply pkg/secrets/vault.go]: Yes, good call. Changing

File: pkg/secrets/vault.go (hunk 2 above)
Comment (KathrynLewis) [reply pkg/secrets/vault.go]: Updated, but not sure how to best check this error...

File: pkg/secrets/vault.go (hunk 3 above)
Comment (KathrynLewis) [reply pkg/secrets/vault.go]: I couldn't find one?

File: pkg/secrets/vault.go [hunk 4]
```diff
@@ -2,23 +2,242 @@ package secrets
...
+	Enabled    bool   `json:"enabled" yaml:"enabled"`
+	Url        string `json:"url" yaml:"url"`
+	AuthMethod string `json:"authMethod" yaml:"authMethod"`
+	Role       string `json:"role" yaml:"role"`
+	Path       string `json:"path" yaml:"path"`
+	Token      string
+}
+
 type VaultSecret struct {
 	engine        string
```
Comment (jasonmcintosh) [inline pkg/secrets/vault.go]: Couldn't this default to secret since that is the default vault sets?

File: pkg/secrets/secrets_test.go [hunk 5]
```diff
@@ -140,3 +141,74 @@ func TestNoSecret(t *testing.T) {
...
+		{
+			"encrypted:vault!e:engine!n:path!k:key!b:true",
+			VaultSecret{
+				engine:   "engine",
+				path:   "path",
+				base64Encoded: "true",
+				key:      "key",
+			},
+			false,
+		},
```
Comment (jasonmcintosh) [inline pkg/secrets/secrets_test.go]: Another test case idea if engine can be defaulted to "secret"
```
	{
			"encrypted:vault!n:path!k:key!b:true",
			VaultSecret{
				engine:   "secret",
				path:   "foo",
				base64Encoded: "true",
				key:      "key",
			},
			false,
		},
```

**Pull request closed without merging**